### PR TITLE
Core Data: Fix redo behavior and expand test coverage.

### LINKED ISSUE #17825: Redoing changes fails
**Describe the bug**

Redoing undone changes fails.

**To reproduce**

1. Type "test".
2. Wait one second or longer.
3. Type again "test"
4.  Undo twice. You should see all both changes undone.
5. Redo twice. Only the first change is redone.

Another case:

1. Type "test".
2. Undo.
3. Redo.
4. Only "t" is restored.

**Expected behavior**

I expect all content to be restored.

**Screenshots**

redo

### PR DESCRIPTION
**Pull request by epiqueras**
Fixes #17825

## Description

This PR fixes an issue where explicit undo level creation was being ignored when the last recorded edit had no non-transient edits, i.e. not the `blocks` property. These explicit undo level actions were also creating an extra undo level, because of some conflicting logic with regular edit actions. This PR also fixes that.

## How has this been tested?

E2E tests were expanded to verify that undoing and redoing creates and uses undo levels correctly.

## Types of Changes

*Bug Fix:* Fix redo levels inconsistencies.

## Checklist:

- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/e2e-tests/specs/undo.test.js [hunk 1]
```diff
@@ -54,27 +54,49 @@ describe( 'undo', () => {
...
-	it( 'Should undo to expected level intervals', async () => {
+	it( 'Should undo/redo to expected level intervals', async () => {
 		await clickBlockAppender();
 
 		await page.keyboard.type( 'This' );
 		await page.keyboard.press( 'Enter' );
 		await page.keyboard.type( 'is' );
 		await page.keyboard.press( 'Enter' );
 		await page.keyboard.type( 'test' );
+		await new Promise( ( resolve ) => setTimeout( resolve, 1000 ) );
```
Comment (ellatrix) [inline packages/e2e-tests/specs/undo.test.js]: Why is this timeout added? Can't we add a separate test case with the timeout? Are we still testing without the timeout if we add it here?

Comment (ellatrix) [reply packages/e2e-tests/specs/undo.test.js]: The test fail when this is removed, which is curious. This line shouldn't be necessary for it to pass.

**Comment by ellatrix** [comment]
I adjusted the test a bit so it also tests if the content is the same. It's still weird that the timeout at the start is needed now, but let's try to resolve that in a separate PR. This is already a vast improvement.

**Review by ellatrix — APPROVED** [review]

File: packages/e2e-tests/specs/undo.test.js (hunk 1 above)
Comment (epiqueras) [reply packages/e2e-tests/specs/undo.test.js]: We need it to make sure the timeout after typing for the undo level creation fires.

**Comment by epiqueras** [comment]
>I adjusted the test a bit so it also tests if the content is the same.

We should bring back the snapshots I added. If we only compare the content against itself like that, we are not actually testing that the levels are being created at the right places, just that the total number of levels is correct. The snapshots actually tested that `// Undo 2nd paragraph text.`, does exactly that and nothing else.

**Comment by ellatrix** [comment]
> We should bring back the snapshots I added. If we only compare the content against itself like that, we are not actually testing that the levels are being created at the right places, just that the total number of levels is correct. The snapshots actually tested that `// Undo 2nd paragraph text.`, does exactly that and nothing else.

It doesn't just test the total number of levels? It compares the content to make sure it's restored correctly. We take snapshots during the creation of the content and compare it to the content produced after undoing and redoing changes.

File: packages/e2e-tests/specs/undo.test.js (hunk 1 above)
Comment (ellatrix) [reply packages/e2e-tests/specs/undo.test.js]: Hm, this shouldn't be necessary. You should be able to undo and redo without waiting a second.

Comment (epiqueras) [reply packages/e2e-tests/specs/undo.test.js]: How does the last level get created otherwise? It's implemented with a timeout.

**Comment by epiqueras** [comment]
Yes, exactly, but it doesn't care about what that content is, so you're just testing the number of levels of content.

**Comment by ellatrix** [comment]
That's not just testing the number of levels? We don't care _what_ the content is, we only care that the content is correctly undone and redone.

File: packages/e2e-tests/specs/undo.test.js (hunk 1 above)
Comment (ellatrix) [reply packages/e2e-tests/specs/undo.test.js]: This is something that was previously tested with `withHistory` unit tests. Do you think it's possible to add unit tests for this new reducer?

**Comment by epiqueras** [comment]
With the snapshots we also test that `// Undo 2nd paragraph text.` is actually undoing the 2nd paragraph text. With these comparisons, we only test that there are a total number of undos and redos and that the state before `// Undo 2nd paragraph text.` is equal to the state after `// Redo 2nd paragraph text.`.

File: packages/e2e-tests/specs/undo.test.js (hunk 1 above)
Comment (epiqueras) [reply packages/e2e-tests/specs/undo.test.js]: Yes

**Comment by ellatrix** [comment]
> and that the state before `// Undo 2nd paragraph text.` is equal to the state after `// Redo 2nd paragraph text.`.

So not just the number. 😉 It also tests that the state is equal to the one before any redo or undo actions occurred. We are saving the content while it's being created:

https://github.com/WordPress/gutenberg/blob/42167a2ee4d5a97c42b1d662a363d0fc35801e2d/packages/e2e-tests/specs/undo.test.js#L60-L80

**Comment by epiqueras** [comment]
Yes, but that code won't detect if the behavior regresses from something like:

`This(Undo Level)is(Undo Level)test`

to

`T(Undo Level)hisi(Undo Level)stest`.

**Comment by ellatrix** [comment]
It would? It would not be the same as the content first created.

**Comment by epiqueras** [comment]
They would be the same, `undo levels === redo levels`, that's why I used snapshots.

**Comment by ellatrix** [comment]
Huh, we are not just comparing the amount of levels...

https://github.com/WordPress/gutenberg/blob/42167a2ee4d5a97c42b1d662a363d0fc35801e2d/packages/e2e-tests/specs/undo.test.js#L98

We are verifying if the content is the correct after undo/redo. If levels would be create in the wrong place, these checks would fail.

**Comment by ellatrix** [comment]
To me this is much clearer than snapshots, as we are explicitly comparing the initially created text with the changes after undo/redo.

**Comment by epiqueras** [comment]
Oh because the levels are created right after every `.type`. I see what you mean. Yeah this makes sense, sorry for the confusion.

File: packages/e2e-tests/specs/undo.test.js (hunk 1 above)
Comment (epiqueras) [reply packages/e2e-tests/specs/undo.test.js]: https://github.com/WordPress/gutenberg/pull/17861/commits/ce90a828cff583b645596cd4b7ee3493741b0468

**Pull request merged**